src/models/oauth_clients.rs

```diff
@@ -0,0 +1,100 @@
+use super::{
+    ids::{Base62Id, UserId},
+    pats::Scopes,
+};
+use chrono::{DateTime, Utc};
+use serde::{Deserialize, Serialize};
+
+#[derive(Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(from = "Base62Id")]
+#[serde(into = "Base62Id")]
+pub struct OAuthClientId(pub u64);
+
+#[derive(Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(from = "Base62Id")]
+#[serde(into = "Base62Id")]
+pub struct OAuthClientAuthorizationId(pub u64);
+
+#[derive(Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(from = "Base62Id")]
+#[serde(into = "Base62Id")]
+pub struct OAuthRedirectUriId(pub u64);
+
+#[derive(Deserialize, Serialize)]
+pub struct OAuthRedirectUri {
+    pub id: OAuthRedirectUriId,
+    pub client_id: OAuthClientId,
+    pub uri: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct OAuthClientCreationResult {
+    pub client_secret: String,
+    pub client: OAuthClient,
+}
```

Comment:
IMO you shouldn't have two separate structs. Just make client_secret an optional field on the client and set it to null by default

Reply:
I disagree and think that's a counterproductive pattern. One of the best things about Rust's expressive type system is that it can be used to make invalid states unrepresentable. This simplifies code, removes possible bugs, and makes it easier to read.

It would be invalid for a client secret to be returned at any time besides during creation. It would also be invalid for a client secret _not_ to be returned when we just created a client. By using the same struct everywhere, both of those invalid scenarios would become possible bugs and need to be guarded against. When they're separate types, those bugs are impossible, and developers don't have to waste any time considering how to handle them.

Reply:
Hmm, that's a fair point but I think you should use serde flatten in this case

Reply:
Great idea, done! Love it.